external/storm-kafka-client/src/main/java/org/apache/storm/kafka/bolt/KafkaBolt.java

```diff
@@ -62,6 +62,7 @@
     private OutputCollector collector;
     private TupleToKafkaMapper<K,V> mapper;
     private KafkaTopicSelector topicSelector;
+    private Callback providedCallback;
```

Comment:
I think we need to define our own interface for this. Callback isn't serializable, so Storm won't be able to transfer it from the Nimbus submitter to the worker JVM.

Reply:
Also I'm not sure exactly what someone implementing this interface would need, but maybe we should add a prepare method to the interface as well, so people who need some configuration or the topology context can get access? What do you think?

Reply:
What about something like this? Not sure if `OutputCollector` is necessary or not.

```java
public interface PreparedCallback extends Callback, Serializable {
    void prepare(Map<String, Object> topoConf, TopologyContext context, OutputCollector outputCollector);
}
```

So if the `PreparedCallback` is not null, then KafkaBolt would call `.prepare()` during `KafkaBolt.prepare()`.

The one concern I have is that it might be open for abuse by tempting people to use it to inject logic into the bolt preparation that's unrelated to Kafka publishing.

Reply:
That looks good to me. I'm also not sure about the collector, but I'd maybe lean toward not including it. The callback won't have access to incoming tuples, so the only use would be if someone wanted the callback to emit a new unanchored tuple.

I wouldn't worry about people misusing the prepare method. They can just as easily override the bolt's prepare method.